miniapps/electromagnetics/maxwell_solver.cpp

```diff
@@ -0,0 +1,673 @@
+   {
+      cout << "Creating discrete curl operator" << endl;
+   }
+   Curl_ = new ParDiscreteCurlOperator(HCurlFESpace_, HDivFESpace_);
+   Curl_->Assemble();
+   Curl_->Finalize();
+   NegCurl_ = Curl_->ParallelAssemble();
+   // Beware this modifies the matrix stored within the Curl_ object.
+   *NegCurl_ *= -1.0;
+
+   // Build grid functions
+   e_    = new ParGridFunction(HCurlFESpace_);
+   dedt_ = new ParGridFunction(HCurlFESpace_);
+   rhs_  = new ParGridFunction(HCurlFESpace_);
+   b_    = new ParGridFunction(HDivFESpace_);
+
+   E_ = e_->ParallelProject();
+   B_ = b_->ParallelProject();
+
+   HD_  = new HypreParVector(HDivFESpace_);
+   RHS_ = new HypreParVector(HCurlFESpace_);
+
+   // Initialize dedt to zero
+   *dedt_ = 0.0;
+
+   if ( j_src_)
+   {
+      if ( myid_ == 0 && logging_ > 0 )
+      {
+         cout << "Creating Current Source" << endl;
+      }
+      jCoef_ = new VectorFunctionCoefficient(3,j_src_);
+
+      j_  = new ParGridFunction(HCurlFESpace_);
+      j_->ProjectCoefficient(*jCoef_);
+
+      jd_ = new ParLinearForm(HCurlFESpace_);
+      jd_->AddDomainIntegrator(new VectorFEDomainLFIntegrator(*jCoef_));
+      jd_->Assemble();
+   }
+
+   dtMax_ = GetMaximumTimeStep();
+}
+
+MaxwellSolver::~MaxwellSolver()
+{
+   delete epsCoef_;
+   delete muInvCoef_;
+   delete etaInvCoef_;
+   delete jCoef_;
+   delete dEdtBCCoef_;
+
+   delete E_;
+   delete B_;
+   delete HD_;
+   delete RHS_;
+
+   delete e_;
+   delete b_;
+   delete j_;
+   delete dedt_;
+   delete rhs_;
+   delete jd_;
+
+   delete Curl_;
+
+   delete M1Losses_;
+   delete M2MuInv_;
+
+   delete hDivMassMuInv_;
+   delete hCurlLosses_;
+   delete weakCurlMuInv_;
+
+   delete HCurlFESpace_;
+   delete HDivFESpace_;
+
+   map<int, ParBilinearForm*>::iterator mit1;
+   for (mit1=a1_.begin(); mit1!=a1_.end(); mit1++)
+   {
+      int i = mit1->first;
+      delete pcg_[i];
+      delete diagScale_[i];
+      delete A1_[i];
+      delete a1_[i];
+   }
+
+   map<int, Coefficient*>::iterator mit2;
+   for (mit2=dtCoef_.begin(); mit2!=dtCoef_.end(); mit2++)
+   {
+      delete mit2->second;
+   }
+   for (mit2=dtSigmaCoef_.begin(); mit2!=dtSigmaCoef_.end(); mit2++)
+   {
+      delete mit2->second;
+   }
+   for (mit2=dtEtaInvCoef_.begin(); mit2!=dtEtaInvCoef_.end(); mit2++)
+   {
+      delete mit2->second;
+   }
+
+   map<string, socketstream*>::iterator mit3;
+   for (mit3=socks_.begin(); mit3!=socks_.end(); mit3++)
+   {
+      delete mit3->second;
+   }
+}
+
+HYPRE_Int
+MaxwellSolver::GetProblemSize()
+{
+   return HCurlFESpace_->GlobalTrueVSize();
+}
+
+void
+MaxwellSolver::PrintSizes()
+{
+   HYPRE_Int size_nd = HCurlFESpace_->GlobalTrueVSize();
+   HYPRE_Int size_rt = HDivFESpace_->GlobalTrueVSize();
+   if ( myid_ == 0 )
+   {
+      cout << "Number of H(Curl) unknowns: " << size_nd << endl;
+      cout << "Number of H(Div)  unknowns: " << size_rt << endl << flush;
+   }
+}
+
+void
+MaxwellSolver::SetInitialEField(VectorCoefficient & EFieldCoef)
```

Comment:
@mlstowell 
I am  hoping this would make more better 
 e_->ProjectCoefficient(*eCoef_);
instead of //e_->ProjectCoefficient(EFieldCoef);

Reply:
Hi, @Asadanees,

These are two different names for the same object so I don't think it would make any difference (unless I've misunderstood your comment).  Admittedly an argument could be made that the member data eCoef_ is never used again so it could be eliminated but that's a different issue.

Thanks,
Mark

Reply:
Hi @mlstowell 
Yes, There is no difference and even did not affect the result. You also answered in the second part that it could be eliminated.  
When you took both the object are same (eCoef_ = &EFieldCoef). than I suggested that it would be more better e_->ProjectCoefficient(*eCoef_). otherwise it could eliminate.
Thanks,
asad anees

Reply:
Perhaps I should remove that member data.  The original idea was to allow the user to pass in a time dependent coefficient describing the exact solution as well as the initial conditions.  The code could then periodically compute the error in its approximate solution.  However, on second thought this behavior should be controlled by the main program which could pass in the coefficient each time it asks for the current error estimate.

Reply:
Hi @mlstowell 
I think, it is more better to take it as it is.  this is more general.
thanks 
asad